File: frontend/types/fundamental.ts
export type FundamentalRating = 'BUY' | 'HOLD' | 'SELL';

export interface FundamentalAnalysisRequest {
  ticker: string;
  market?: 'US' | 'IN';
  include_peer_context?: boolean;
}

export interface FundamentalPeerContext {
  sector_percentile: number | null;
  universe_percentile: number | null;
  relative_rank: number | null;
  source: string;
}

export interface FundamentalAnalysisResponse {
  ticker: string;
  market: string;
  company_name: string;
  sector: string | null;
  rating: FundamentalRating;
  signal: string;
  score: number;
  model_score: number | null;
  universe_percentile: number | null;
  relative_rank: number | null;
  key_metrics: {
    pe_ratio: number | null;
    roe: number | null;
    debt_to_equity: number | null;
    free_cash_flow_margin: number | null;
    revenue_growth_yoy: number | null;
    earnings_growth_yoy: number | null;
    [key: string]: number | null;
  };
  trends: Record<string, string>;
  peer_context: FundamentalPeerContext | null;
  strengths: string[];
  concerns: string[];
  analysis_summary: string;
  data_source: string;
  cached: boolean;
  source_signal_date: string | null;
  generated_at: string;
}

File: backend/app/schemas/fundamental.py
from datetime import datetime
from typing import Dict, List, Literal, Optional

from pydantic import BaseModel, Field


class FundamentalAnalysisRequest(BaseModel):
    ticker: str = Field(..., min_length=1, max_length=20, description="Stock ticker symbol")
    market: Literal["US", "IN"] = Field("US", description="Market: US or India")
    include_peer_context: bool = Field(True, description="Include ranking/peer context when available")


class FundamentalPeerContext(BaseModel):
    sector_percentile: Optional[float] = Field(None, ge=0.0, le=1.0)
    universe_percentile: Optional[float] = Field(None, ge=0.0, le=1.0)
    relative_rank: Optional[int] = Field(None, ge=1)
    source: str


class FundamentalAnalysisResponse(BaseModel):
    ticker: str
    market: str
    company_name: str
    sector: Optional[str] = None
    rating: Literal["BUY", "HOLD", "SELL"]
    signal: str
    score: float = Field(..., ge=0.0, le=10.0)
    model_score: Optional[float] = Field(None, ge=0.0, le=1.0)
    universe_percentile: Optional[float] = Field(None, ge=0.0, le=1.0)
    relative_rank: Optional[int] = Field(None, ge=1)
    key_metrics: Dict[str, Optional[float]]
    trends: Dict[str, str]
    peer_context: Optional[FundamentalPeerContext] = None
    strengths: List[str]
    concerns: List[str]
    analysis_summary: str
    data_source: str
    cached: bool
    source_signal_date: Optional[str] = None
    generated_at: datetime
